libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,183 @@
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_TempSensorCmd(ENABLE);
+  ADC_VrefintCmd(ENABLE);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, !adc_mode);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+  ADC_ITConfig(ADC1, ADC_IER_EOSEQIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.sequence = ADC1->CHSELR;
+
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+
+  // Set reference voltage channel on by default
+  ADC_ChannelConfig(ADC1, (1 << ADC_CHANNEL_REF), ADC_SampleTime_239_5Cycles);
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL || adc_channel == ADC_CHANNEL_REF) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= ~(1 << adc_channel);
+  }
+
+  // The bridge divider is only enabled when needed to minimize consumption on the battery
+  if (adc_channel == ADC_CHANNEL_BAT) {
+    ADC_VbatCmd(new_state);
+  }
+
+  s_adc_status.sequence = ADC1->CHSELR;
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(ADCChannel adc_channel, ADCCallback callback, void *context) {
+  // Returns invalid if the given address is not connected to an ADC channel
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
+  s_adc_interrupts[adc_channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_read_value(ADCChannel adc_channel, uint16_t *reading) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (!s_adc_status.continuous) {
+    ADC_StartOfConversion(ADC1);
+    while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADSTART)) { }
+  }
+
+  *reading = s_adc_interrupts[adc_channel].reading;
+
+  switch (adc_channel) {
+    case ADC_CHANNEL_TEMP:
+      *reading = prv_get_temp(*reading);
+      break;
+
+    case ADC_CHANNEL_REF:
+      *reading = prv_get_vdda(*reading);
+      break;
+
+    case ADC_CHANNEL_BAT:
+      *reading *= 2;
+      break;
+
+    default:
+      break;
+  }
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_read_voltage(ADCChannel adc_channel, uint16_t *reading) {
```

Comment:
It might make more sense to make this a "read converted" type of function that returns temperature (in K?) for the temp sensor and converted voltage for everything else.